File: codeForces/C_2_Game_on_Tree_Medium.py
import sys
input = sys.stdin.read

def dfs(nodo, padre, h):
    hoja = True
    tM = True
    tB = True

    for u in g[nodo]:
        if padre == u:
            continue

        hoja = False
        dfs(u, nodo, h + 1)

        if not hoja:
            tM &= not win[u]
            tB &= win[u]

    win[nodo] = (hoja and (h % 2 == 0))

    if not hoja:
        if (h & 1 and not tM) or (h % 2 == 0 and tB):
            win[nodo] = True


data = input().split()
index = 0

n = int(data[index])
index += 1
t = int(data[index])
index += 1

g = [[] for _ in range(n + 1)]
win = [0] * (n + 1)

for _ in range(n - 1):
    u = int(data[index])
    index += 1
    v = int(data[index])
    index += 1
    g[u].append(v)
    g[v].append(u)

raiz = int(data[index])

dfs(raiz, 0, 0)
print("Hermione" if win[raiz] else "Ron")

File: codeForces/C_2_Game_on_Tree_Medium.cpp
#include<bits/stdc++.h>
using namespace std;

vector<vector<int>> g;
vector<int> win;
int raiz;

void dfs (int nodo, int padre, int h) {
    bool hoja = true, tM = true, tB = true;

    for (auto u: g[nodo]) {
        if (padre == u) continue;

        hoja = false;
        dfs(u, nodo, h + 1);

        if (!hoja) {
            tM &= !win[u];
            tB &= win[u];
        }   
    }

    win[nodo] = (hoja && (h % 2 == 0));

    if (!hoja) {
        if ((h & 1 && !tM) || (h % 2 == 0 && tB)) {
            win[nodo] = true;
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    int n, t;
    cin >> n >> t;

    g = vector<vector<int>>(n + 1, vector<int>());
    win = vector<int>(n + 1, 0);

    for (int i = 0; i < n - 1; i++) {
        int u, v;
        cin >> u >> v;
        g[u].push_back(v);
        g[v].push_back(u);
    }

    cin >> raiz;
    
    dfs(raiz, 0, 0);
    cout << (win[raiz] ? "Hermione" : "Ron") << "\n";

    return 0;
}